### PR TITLE
Implement ekfac with new interface

### LINKED ISSUE #451: Implement EKFAC
First implementation. See #452 for a continuation

### PR DESCRIPTION
**Pull request by Xuzzo**
### Description

This PR closes #451 

### Changes

- Adds new method for influences: Ekfac
- adds related documentation and tests
- Adds section in wine notebook to showcase use of new methods

### Checklist

- [x] Wrote Unit tests (if necessary)
- [x] Updated Documentation (if necessary)
- [x] Updated Changelog
- [x] If notebooks were added/changed, added boilerplate cells are tagged with `"tags": ["hide"]` or `"tags": ["hide-input"]`

### CODE REVIEW

File: notebooks/influence_wine.ipynb [hunk 1]
```diff
@@ -767,6 +775,236 @@
...
+      ]
+     },
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "plt.plot(mean_ekfac_train_influences, color=\"red\")\n",
```
Comment (schroedk) [inline notebooks/influence_wine.ipynb:916]: My suggestion would be to use a scatter plot instead, so one could see the correlation immediately.

File: src/pydvl/influence/torch/influence_function_model.py [hunk 2]
```diff
@@ -876,3 +880,321 @@ def to(self, device: torch.device):
...
+    factorization of the Hessian first introduced in [@martens2015optimizing].
+    Contrary to the other influence function methods, this implementation can only
+    be used for classification tasks with a cross entropy loss function. However, it
+    is much faster than the other methods and can be used efficiently for very large
+    datasets and models. For more information, see [Eigenvalue Corrected K-FAC][ekfac].
+
+    Args:
+        model: Instance of [torch.nn.Module][torch.nn.Module].
+        hessian_regularization: Regularization of the hessian.
+    """
```
Comment (schroedk) [inline src/pydvl/influence/torch/influence_function_model.py:899]: My suggestion would be to make all the methods related to the details of ekfac invisible in the documentation, i.e. add an underscore (make them "private").
If you really want to expose them to the user, my suggestion would be to make the docstrings complete, i.e. with input and output description in google format.

File: src/pydvl/influence/torch/util.py [hunk 3]
```diff
@@ -453,3 +454,57 @@ def __call__(
...
+            [evec_a.to(device) for evec_a in self.evecs_a],
+            [evec_g.to(device) for evec_g in self.evecs_g],
+            [diag.to(device) for diag in self.diags],
+        )
+
+
+def empirical_cross_entropy_loss_fn(
+    model_output: torch.Tensor, *args, **kwargs
+) -> torch.Tensor:
+    probs_ = torch.softmax(model_output, dim=1)
```
Comment (schroedk) [inline src/pydvl/influence/torch/util.py:505]: Please add a docstring to this function, maybe explaining the *args and **kwargs?

**Review by schroedk — CHANGES_REQUESTED** [review]

File: docs/assets/pydvl.bib [hunk 4]
```diff
@@ -342,4 +342,21 @@ @InProceedings{kwon_data_2023
...
+}
+
+@article{george2018fast,
```
[the remark this answers is not in the document]
Comment (Xuzzo) [reply docs/assets/pydvl.bib:347]: I'll add it in the notebook. It does not add much from the theory perspective, it is just focussed on LLMs

File: docs/influence/influence_function_model.md [hunk 5]
```diff
@@ -97,4 +97,32 @@ if_model = ArnoldiInfluence
...
+A further improvement of the K-FAC method is the Eigenvalue Corrected K-FAC (EKFAC) method [@george2018fast], which allows to further re-fit the eigenvalues of the Hessian, thus providing a more accurate approximation. On top of the K-FAC method, the EKFAC method is implemented by simply calling the update_diag method from [EkfacInfluence](pydvl/influence/torch/influence_function_model.py). The following code snippet shows how to use the EKFAC method to calculate the influence function of a model. 
+
+```python
+from pydvl.influence.torch import EkfacInfluence
+if_model = EkfacInfluence(
+    model,
+    hessian_regularization=0.0,
+)
+if_model.fit(train_loader)
+if_model.update_diag(train_loader)
```
[the remark this answers is not in the document]
Comment (Xuzzo) [reply docs/influence/influence_function_model.md:127]: No updating the diagonal is optional. In most cases it improves things quite little and it takes as much time as fitting the k-fac blocks.

File: src/pydvl/influence/torch/util.py [hunk 6]
```diff
@@ -453,3 +454,57 @@ def __call__(
...
+    def __iter__(self):
+        return iter(
+            zip(
+                self.layer_names,
+                zip(self.layers_module, self.evecs_a, self.evecs_g, self.diags),
+            )
+        )
+
+    def get_layer_evecs(self):
+        evecs_a_dict = {layer_name: evec_a for layer_name, (_, evec_a, _, _) in self}
```
[the remark this answers is not in the document]
Comment (Xuzzo) [reply src/pydvl/influence/torch/util.py:495]: I would keep it public. I have added some docs. Let me know if more needed

File: src/pydvl/influence/torch/influence_function_model.py [hunk 7]
```diff
@@ -876,3 +881,322 @@ def to(self, device: torch.device):
...
+                ).view(-1)
+
+        else:
+            raise NotImplementedLayerRepresentationException(
+                f"Only Linear layers are supported, but found module {module} requiring grad."
+            )
+        return input_hook, grad_hook
+
+    def update_diag(
+        self,
```
Comment (schroedk) [inline src/pydvl/influence/torch/influence_function_model.py:1118]: Still, having this separate method could be avoided. Is my understanding correct, that this is the eigenvalue correction (so the E in Ekfac?). I would suggest to introduce an init parameter, if the kfac blocks should be adapted to this update logic. What do you think?

**Comment by mdbenito** [comment]
@Xuzzo Remember to add the paper to the readme!

**Pull request merged**